File: java/Samples/Biometrics/CS/ServerSampleCS/Connections/ConnectionForm.cs
using System;
using System.IO;
using System.Windows.Forms;
using Neurotec.Samples.Code;

namespace Neurotec.Samples.Connections
{
	public partial class ConnectionForm : Form
	{
		#region Public constructor

		public ConnectionForm()
		{
			InitializeComponent();
			LoadSettings();
		}

		#endregion

		#region Private fields

		private DatabaseConnection _db;

		#endregion Private fields

		#region Public properties

		public string Server
		{
			get { return tbMMAServer.Text; }
			set { tbMMAServer.Text = value; }
		}

		public int ClientPort
		{
			get { return (int)nudPort.Value; }
			set { nudPort.Value = value; }
		}

		public int AdminPort
		{
			get { return (int)nudAdminPort.Value; }
			set { nudAdminPort.Value = value; }
		}

		public string Password
		{
			get { return mtbPasword.Text; }
			set { mtbPasword.Text = value; }
		}

		public string UserName
		{
			get { return tbUsername.Text; }
			set { tbUsername.Text = value; }
		}

		public bool UseDb
		{
			get { return rbDatabase.Checked; }
			set
			{
				if (value) rbDatabase.Checked = true;
				else rbDirectory.Checked = true;
				ToggleRadioButtons();
			}
		}

		public string TemplateDir
		{
			get { return tbPath.Text; }
			set { tbPath.Text = value; }
		}

		public string DbServer
		{
			get { return tbDBServer.Text; }
			set { tbDBServer.Text = value; }
		}

		public string DbTable
		{
			get { return cbTable.Text; }
		}

		public string DbUser
		{
			get { return tbDBUser.Text; }
			set { tbDBUser.Text = value; }
		}

		public string DbPassword
		{
			get { return tbDBPassword.Text; }
			set { tbDBPassword.Text = value; }
		}

		public bool IsAccelerator
		{
			get { return chbIsAccelerator.Checked; }
			set { chbIsAccelerator.Checked = value; }
		}

		public DatabaseConnection Db
		{
			get { return _db; }
			set
			{
				_db = value;
				if (value == null) return;

				DbServer = value.Server;
				DbUser = value.User;
				DbPassword = value.Password;
				cbTable.Items.Clear();
				cbTable.Items.AddRange(value.GetTables());
				cbTable.SelectedItem = value.Table;
				ListCollumns();
				cbId.SelectedItem = value.IdColumn;
				cbTemplate.SelectedItem = value.TemplateColumn;
			}
		}

		#endregion

		#region Private methods

		private void LoadSettings()
		{
			UseDb = SettingsAccesor.UseDb;
			DbServer = SettingsAccesor.DbServer;
			DbUser = SettingsAccesor.DbUser;
			DbPassword = SettingsAccesor.DbPassword;
			Server = SettingsAccesor.Server;
			ClientPort = SettingsAccesor.ClientPort;
			AdminPort = SettingsAccesor.AdminPort;
			UserName = SettingsAccesor.UserName;
			Password = SettingsAccesor.Password;
			TemplateDir = SettingsAccesor.TemplateDir;
			IsAccelerator = SettingsAccesor.IsAccelerator;
		}

		private void ToggleRadioButtons()
		{
			gbDatabase.Enabled = rbDatabase.Checked;
			btnBrowse.Enabled = tbPath.Enabled = rbDirectory.Checked;
		}

		private void SaveSettings()
		{
			SettingsAccesor.UseDb = UseDb;
			SettingsAccesor.DbServer = DbServer;
			string value = DbTable;
			if (!string.IsNullOrEmpty(value))
				SettingsAccesor.DbTable = value;
			SettingsAccesor.DbUser = DbUser;
			SettingsAccesor.DbPassword = DbPassword;
			value = cbId.SelectedItem as string;
			if (!string.IsNullOrEmpty(value))
				SettingsAccesor.IdColumn = value;
			value = cbTemplate.SelectedItem as string;
			if (!string.IsNullOrEmpty(value))
				SettingsAccesor.TemplateColumn = value;
			SettingsAccesor.Password = Password;
			SettingsAccesor.UserName = UserName;
			SettingsAccesor.ClientPort = ClientPort;
			SettingsAccesor.AdminPort = AdminPort;
			SettingsAccesor.Server = Server;
			SettingsAccesor.TemplateDir = TemplateDir;
			SettingsAccesor.IsAccelerator = IsAccelerator;
			SettingsAccesor.SaveSettings();
		}

		private void DatabaseSettingsReset()
		{
			Db = null;
			tbDBServer.Text = @"mysql_dsn";
			tbDBUser.Text = @"user";
			tbDBPassword.Text = @"pass";
			try
			{
				var conn = new DatabaseConnection();
				string hostValue = conn.GetConfigValue("Host");
				if (hostValue != null)
				{
					tbDBServer.Text = hostValue;
				}
				string userValue = conn.GetConfigValue("User");
				if (userValue != null)
				{
					tbDBUser.Text = userValue;
				}
				string passwordValue = conn.GetConfigValue("Password");
				if (passwordValue != null)
				{
					tbDBPassword.Text = passwordValue;
				}
			}
			catch { }
			cbTable.Items.Clear();
			cbTemplate.Items.Clear();
			cbId.Items.Clear();
		}

		private void ListCollumns()
		{
			cbId.Items.Clear();
			cbTemplate.Items.Clear();
			if (cbTable.SelectedItem != null)
			{
				string[] columns = Db.GetColumns(DbTable);
				if (columns != null)
				{
					cbId.Items.AddRange(columns);
					cbTemplate.Items.AddRange(columns);
				}
			}
		}

		#endregion

		#region Private form events

		private void BtnConnectClick(object sender, EventArgs e)
		{
			if (Db != null)
			{
				if (Db.Server == DbServer && Db.User == DbUser &&
					Db.Password == DbPassword)
					return;
			}
			_db = new DatabaseConnection();

			try
			{
				if (_db == null) throw new ApplicationException("Specified db is null");

				_db.Server = DbServer;
				_db.User = DbUser;
				_db.Password = DbPassword;
				_db.CheckConnection();
				cbTable.Items.Clear();
				string[] tables = _db.GetTables();
				if (tables != null)
				{
					cbTable.Items.AddRange(tables);
					string table = SettingsAccesor.DbTable;
					if (cbTable.Items.Contains(table))
					{
						cbTable.SelectedItem = table;
					}
					else if (tables.Length > 0)
						cbTable.SelectedIndex = 0;
				}
			}
			catch (Exception ex)
			{
				_db = null;
				Utilities.ShowError("Failed to connect to db. Reason: {0}", ex);
			}
		}

		private void CbTableSelectedIndexChanged(object sender, EventArgs e)
		{
			if (Db != null) Db.Table = cbTable.Text;
			ListCollumns();

			string collumn = SettingsAccesor.IdColumn;
			if (cbId.Items.Contains(collumn))
				cbId.SelectedItem = collumn;
			else if (cbId.Items.Count > 0)
				cbId.SelectedIndex = 0;

			collumn = SettingsAccesor.TemplateColumn;
			if (cbTemplate.Items.Contains(collumn))
				cbTemplate.SelectedItem = collumn;
			else if (cbTemplate.Items.Count > 0)
				cbTemplate.SelectedIndex = 0;
		}

		private void BtnResetClick(object sender, EventArgs e)
		{
			DatabaseSettingsReset();
		}

		private void ChbIsAcceleratorCheckedChanged(object sender, EventArgs e)
		{
			tbUsername.Enabled = chbIsAccelerator.Checked;
			mtbPasword.Enabled = chbIsAccelerator.Checked;
		}

		private void BtnCheckConnectionClick(object sender, EventArgs e)
		{
			if (AcceleratorConnection.CheckConnection(tbMMAServer.Text, (int)nudAdminPort.Value))
			{
				Utilities.ShowInformation("Connection test successful");
			}
			else
			{
				Utilities.ShowError("Connection failed");
			}
		}

		private void BtnBrowseClick(object sender, EventArgs e)
		{
			string path = TemplateDir;
			if (!string.IsNullOrEmpty(path) && Directory.Exists(path))
				folderBrowserDialog.SelectedPath = path;
			if (folderBrowserDialog.ShowDialog(this) == DialogResult.OK)
				TemplateDir = folderBrowserDialog.SelectedPath;
		}

		private void RbDirectoryCheckedChanged(object sender, EventArgs e)
		{
			ToggleRadioButtons();
		}

		private void BtnOkClick(object sender, EventArgs e)
		{
			if (UseDb)
			{
				if (Db == null)
				{
					Utilities.ShowInformation("Connection with database must be established before proceeding");
					return;
				}
				Db.Table = cbTable.Text;
				Db.TemplateColumn = cbTemplate.Text;
				Db.IdColumn = cbId.Text;
			}
			else if (string.IsNullOrEmpty(TemplateDir) || !Directory.Exists(TemplateDir))
			{
				Utilities.ShowInformation("Specified directory doesn't exists");
				return;
			}

			SaveSettings();
			DialogResult = DialogResult.OK;
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/ServerSampleVB/Connections/ConnectionForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Windows.Forms

Partial Public Class ConnectionForm
	Inherits Form
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
		LoadSettings()
	End Sub

#End Region

#Region "Private fields"

	Private _db As DatabaseConnection

#End Region	' Private fields

#Region "Public properties"

	Public Property Server() As String
		Get
			Return tbMMAServer.Text
		End Get
		Set(ByVal value As String)
			tbMMAServer.Text = value
		End Set
	End Property

	Public Property ClientPort() As Integer
		Get
			Return CInt(Fix(nudPort.Value))
		End Get
		Set(ByVal value As Integer)
			nudPort.Value = value
		End Set
	End Property

	Public Property AdminPort() As Integer
		Get
			Return CInt(Fix(nudAdminPort.Value))
		End Get
		Set(ByVal value As Integer)
			nudAdminPort.Value = value
		End Set
	End Property

	Public Property Password() As String
		Get
			Return mtbPasword.Text
		End Get
		Set(ByVal value As String)
			mtbPasword.Text = value
		End Set
	End Property

	Public Property UserName() As String
		Get
			Return tbUsername.Text
		End Get
		Set(ByVal value As String)
			tbUsername.Text = value
		End Set
	End Property

	Public Property UseDb() As Boolean
		Get
			Return rbDatabase.Checked
		End Get
		Set(ByVal value As Boolean)
			If value Then
				rbDatabase.Checked = True
			Else
				rbDirectory.Checked = True
			End If
			ToggleRadioButtons()
		End Set
	End Property

	Public Property TemplateDir() As String
		Get
			Return tbPath.Text
		End Get
		Set(ByVal value As String)
			tbPath.Text = value
		End Set
	End Property

	Public Property DbServer() As String
		Get
			Return tbDBServer.Text
		End Get
		Set(ByVal value As String)
			tbDBServer.Text = value
		End Set
	End Property

	Public ReadOnly Property DbTable() As String
		Get
			Return cbTable.Text
		End Get
	End Property

	Public Property DbUser() As String
		Get
			Return tbDBUser.Text
		End Get
		Set(ByVal value As String)
			tbDBUser.Text = value
		End Set
	End Property

	Public Property DbPassword() As String
		Get
			Return tbDBPassword.Text
		End Get
		Set(ByVal value As String)
			tbDBPassword.Text = value
		End Set
	End Property

	Public Property IsAccelerator() As Boolean
		Get
			Return chbIsAccelerator.Checked
		End Get
		Set(ByVal value As Boolean)
			chbIsAccelerator.Checked = value
		End Set
	End Property

	Public Property Db() As DatabaseConnection
		Get
			Return _db
		End Get
		Set(ByVal value As DatabaseConnection)
			_db = value
			If value Is Nothing Then
				Return
			End If

			DbServer = value.Server
			DbUser = value.User
			DbPassword = value.Password
			cbTable.Items.Clear()
			cbTable.Items.AddRange(value.GetTables())
			cbTable.SelectedItem = value.Table
			ListCollumns()
			cbId.SelectedItem = value.IdColumn
			cbTemplate.SelectedItem = value.TemplateColumn
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub LoadSettings()
		UseDb = SettingsAccesor.UseDb
		DbServer = SettingsAccesor.DbServer
		DbUser = SettingsAccesor.DbUser
		DbPassword = SettingsAccesor.DbPassword
		Server = SettingsAccesor.Server
		ClientPort = SettingsAccesor.ClientPort
		AdminPort = SettingsAccesor.AdminPort
		UserName = SettingsAccesor.UserName
		Password = SettingsAccesor.Password
		TemplateDir = SettingsAccesor.TemplateDir
		IsAccelerator = SettingsAccesor.IsAccelerator
	End Sub

	Private Sub ToggleRadioButtons()
		gbDatabase.Enabled = rbDatabase.Checked
		tbPath.Enabled = rbDirectory.Checked
		btnBrowse.Enabled = tbPath.Enabled
	End Sub

	Private Sub SaveSettings()
		SettingsAccesor.UseDb = UseDb
		SettingsAccesor.DbServer = DbServer
		Dim value As String = DbTable
		If (Not String.IsNullOrEmpty(value)) Then
			SettingsAccesor.DbTable = value
		End If
		SettingsAccesor.DbUser = DbUser
		SettingsAccesor.DbPassword = DbPassword
		value = TryCast(cbId.SelectedItem, String)
		If (Not String.IsNullOrEmpty(value)) Then
			SettingsAccesor.IdColumn = value
		End If
		value = TryCast(cbTemplate.SelectedItem, String)
		If (Not String.IsNullOrEmpty(value)) Then
			SettingsAccesor.TemplateColumn = value
		End If
		SettingsAccesor.Password = Password
		SettingsAccesor.UserName = UserName
		SettingsAccesor.ClientPort = ClientPort
		SettingsAccesor.AdminPort = AdminPort
		SettingsAccesor.Server = Server
		SettingsAccesor.TemplateDir = TemplateDir
		SettingsAccesor.IsAccelerator = IsAccelerator
		SettingsAccesor.SaveSettings()
	End Sub

	Private Sub DatabaseSettingsReset()
		Db = Nothing
		tbDBServer.Text = "mysql_dsn"
		tbDBUser.Text = "user"
		tbDBPassword.Text = "pass"
		Try
			Dim conn = New DatabaseConnection()
			Dim hostValue As String = conn.GetConfigValue("Host")
			If hostValue IsNot Nothing Then
				tbDBServer.Text = hostValue
			End If
			Dim userValue As String = conn.GetConfigValue("User")
			If userValue IsNot Nothing Then
				tbDBUser.Text = userValue
			End If
			Dim passwordValue As String = conn.GetConfigValue("Password")
			If passwordValue IsNot Nothing Then
				tbDBPassword.Text = passwordValue
			End If
		Catch
		End Try
		cbTable.Items.Clear()
		cbTemplate.Items.Clear()
		cbId.Items.Clear()
	End Sub

	Private Sub ListCollumns()
		cbId.Items.Clear()
		cbTemplate.Items.Clear()
		If cbTable.SelectedItem IsNot Nothing Then
			Dim columns() As String = Db.GetColumns(DbTable)
			If columns IsNot Nothing Then
				cbId.Items.AddRange(columns)
				cbTemplate.Items.AddRange(columns)
			End If
		End If
	End Sub

#End Region

#Region "Private form events"

	Private Sub BtnConnectClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnConnect.Click
		If Db IsNot Nothing Then
			If Db.Server = DbServer AndAlso Db.User = DbUser AndAlso Db.Password = DbPassword Then
				Return
			End If
		End If
		_db = New DatabaseConnection()

		Try
			If _db Is Nothing Then
				Throw New ApplicationException("Specified db is null")
			End If

			_db.Server = DbServer
			_db.User = DbUser
			_db.Password = DbPassword
			_db.CheckConnection()
			cbTable.Items.Clear()
			Dim tables() As String = _db.GetTables()
			If tables IsNot Nothing Then
				cbTable.Items.AddRange(tables)
				Dim table As String = SettingsAccesor.DbTable
				If cbTable.Items.Contains(table) Then
					cbTable.SelectedItem = table
				ElseIf tables.Length > 0 Then
					cbTable.SelectedIndex = 0
				End If
			End If
		Catch ex As Exception
			_db = Nothing
			Utilities.ShowError("Failed to connect to db. Reason: {0}", ex)
		End Try
	End Sub

	Private Sub CbTableSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbTable.SelectedIndexChanged
		If Db IsNot Nothing Then
			Db.Table = cbTable.Text
		End If
		ListCollumns()

		Dim collumn As String = SettingsAccesor.IdColumn
		If cbId.Items.Contains(collumn) Then
			cbId.SelectedItem = collumn
		ElseIf cbId.Items.Count > 0 Then
			cbId.SelectedIndex = 0
		End If

		collumn = SettingsAccesor.TemplateColumn
		If cbTemplate.Items.Contains(collumn) Then
			cbTemplate.SelectedItem = collumn
		ElseIf cbTemplate.Items.Count > 0 Then
			cbTemplate.SelectedIndex = 0
		End If
	End Sub

	Private Sub BtnResetClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnReset.Click
		DatabaseSettingsReset()
	End Sub

	Private Sub ChbIsAcceleratorCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbIsAccelerator.CheckedChanged
		tbUsername.Enabled = chbIsAccelerator.Checked
		mtbPasword.Enabled = chbIsAccelerator.Checked
	End Sub

	Private Sub BtnCheckConnectionClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnCheckConnection.Click
		If AcceleratorConnection.CheckConnection(tbMMAServer.Text, CInt(Fix(nudAdminPort.Value))) Then
			Utilities.ShowInformation("Connection test successful")
		Else
			Utilities.ShowError("Connection failed")
		End If
	End Sub

	Private Sub BtnBrowseClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnBrowse.Click
		Dim path As String = TemplateDir
		If (Not String.IsNullOrEmpty(path)) AndAlso Directory.Exists(path) Then
			folderBrowserDialog.SelectedPath = path
		End If
		If folderBrowserDialog.ShowDialog(Me) = System.Windows.Forms.DialogResult.OK Then
			TemplateDir = folderBrowserDialog.SelectedPath
		End If
	End Sub

	Private Sub RbDirectoryCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles rbDirectory.CheckedChanged
		ToggleRadioButtons()
	End Sub

	Private Sub BtnOkClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOk.Click
		If UseDb Then
			If Db Is Nothing Then
				Utilities.ShowInformation("Connection with database must be established before proceeding")
				Return
			End If
			Db.Table = cbTable.Text
			Db.TemplateColumn = cbTemplate.Text
			Db.IdColumn = cbId.Text
		ElseIf String.IsNullOrEmpty(TemplateDir) OrElse (Not Directory.Exists(TemplateDir)) Then
			Utilities.ShowInformation("Specified directory doesn't exists")
			Return
		End If

		SaveSettings()
		DialogResult = System.Windows.Forms.DialogResult.OK
	End Sub

#End Region
End Class
